electrum/gui/qt/wizard/wallet.py

```diff
@@ -0,0 +1,1384 @@
+    scanComplete = pyqtSignal()
+
+    def __init__(self, parent, wizard):
+        WizardComponent.__init__(self, parent, wizard, title=_('Choose Hardware Device'))
+        Logger.__init__(self)
+        self.scanFailed.connect(self.on_scan_failed)
+        self.scanComplete.connect(self.on_scan_complete)
+        self.plugins = wizard.plugins
+
+        self.error_l = WWLabel()
+        self.error_l.setVisible(False)
+
+        self.device_list = QWidget()
+        self.device_list_layout = QVBoxLayout()
+        self.device_list.setLayout(self.device_list_layout)
+        self.choice_w = None
+
+        self.rescan_button = QPushButton(_('Rescan devices'))
+        self.rescan_button.clicked.connect(self.on_rescan)
+
+        hbox = QHBoxLayout()
+        hbox.addStretch(1)
+        hbox.addWidget(self.rescan_button)
+        hbox.addStretch(1)
+
+        self.layout().addWidget(self.error_l)
+        self.layout().addWidget(self.device_list)
+        self.layout().addStretch(1)
+        self.layout().addLayout(hbox)
+        self.layout().addStretch(1)
+
+        self.c_values = []
+
+    def on_ready(self):
+        self.scan_devices()
+
+    def on_rescan(self):
+        self.scan_devices()
+
+    def on_scan_failed(self, code, message):
+        self.error_l.setText(message)
+        self.error_l.setVisible(True)
+        self.device_list.setVisible(False)
+
+        self.valid = False
+
+    def on_scan_complete(self):
+        self.error_l.setVisible(False)
+        self.device_list.setVisible(True)
+
+        choices = []
+        for name, info in self.devices:
+            state = _("initialized") if info.initialized else _("wiped")
+            label = info.label or _("An unnamed {}").format(name)
+            try:
+                transport_str = info.device.transport_ui_string[:20]
+            except Exception:
+                transport_str = 'unknown transport'
+            descr = f"{label} [{info.model_name or name}, {state}, {transport_str}]"
+            choices.append(((name, info), descr))
+        msg = _('Select a device') + ':'
+
+        # remove old component before adding anew
+        a = self.device_list.layout().itemAt(0)
+        self.device_list.layout().removeItem(a)
+
+        self.choice_w = ChoiceWidget(message=msg, choices=choices)
+        self.device_list.layout().addWidget(self.choice_w)
+
+        self.valid = True
+
+        if self.valid:
+            self.wizard.next_button.setFocus()
+        else:
+            self.rescan_button.setFocus()
+
+
+    def failed_getting_device_infos(self, debug_msg, name, e):
+        # nonlocal debug_msg
+        err_str_oneline = ' // '.join(str(e).splitlines())
+        self.logger.warning(f'error getting device infos for {name}: {err_str_oneline}')
+        indented_error_msg = '    '.join([''] + str(e).splitlines(keepends=True))
+        debug_msg += f'  {name}: (error getting device infos)\n{indented_error_msg}\n'
```

Comment:
debug_msg is not getting side-effected correctly. Only the logging is visible to the outside atm.
(Just insert some artificial exceptions somewhere to test. -- I saw some organically)
Maybe something like this could work:
```patch
diff --git a/electrum/gui/qt/wizard/wallet.py b/electrum/gui/qt/wizard/wallet.py
index 6169f9d618..016e4f1184 100644
--- a/electrum/gui/qt/wizard/wallet.py
+++ b/electrum/gui/qt/wizard/wallet.py
@@ -1134,12 +1134,7 @@ class WCChooseHWDevice(WizardComponent, Logger):
             self.rescan_button.setFocus()
 
 
-    def failed_getting_device_infos(self, debug_msg, name, e):
-        # nonlocal debug_msg
-        err_str_oneline = ' // '.join(str(e).splitlines())
-        self.logger.warning(f'error getting device infos for {name}: {err_str_oneline}')
-        indented_error_msg = '    '.join([''] + str(e).splitlines(keepends=True))
-        debug_msg += f'  {name}: (error getting device infos)\n{indented_error_msg}\n'
+
 
     def scan_devices(self):
         self.valid = False
@@ -1153,6 +1148,13 @@ class WCChooseHWDevice(WizardComponent, Logger):
             devmgr = self.plugins.device_manager
             debug_msg = ''
 
+            def failed_getting_device_infos(name, e):
+                nonlocal debug_msg
+                err_str_oneline = ' // '.join(str(e).splitlines())
+                self.logger.warning(f'error getting device infos for {name}: {err_str_oneline}')
+                indented_error_msg = '    '.join([''] + str(e).splitlines(keepends=True))
+                debug_msg += f'  {name}: (error getting device infos)\n{indented_error_msg}\n'
+
             # scan devices
             try:
                 # scanned_devices = self.run_task_without_blocking_gui(task=devmgr.scan_devices,
@@ -1178,15 +1180,15 @@ class WCChooseHWDevice(WizardComponent, Logger):
                         device_infos = devmgr.list_pairable_device_infos(
                             handler=None, plugin=plugin, devices=scanned_devices, include_failing_clients=True)
                     except HardwarePluginLibraryUnavailable as e:
-                        self.failed_getting_device_infos(debug_msg, name, e)
+                        failed_getting_device_infos(name, e)
                         continue
                     except BaseException as e:
                         self.logger.exception('')
-                        self.failed_getting_device_infos(debug_msg, name, e)
+                        failed_getting_device_infos(name, e)
                         continue
                     device_infos_failing = list(filter(lambda di: di.exception is not None, device_infos))
                     for di in device_infos_failing:
-                        self.failed_getting_device_infos(debug_msg, name, di.exception)
+                        failed_getting_device_infos(name, di.exception)
                     device_infos_working = list(filter(lambda di: di.exception is None, device_infos))
                     devices += list(map(lambda x: (name, x), device_infos_working))
             if not debug_msg:
```

Reply:
Thanks, committed.